### PR TITLE
[CB] Python Speculative decoding

### PR DESCRIPTION
**Pull request by iefode**
Merge after: [PR907](https://github.com/openvinotoolkit/openvino.genai/pull/907/files) to enable speculative decoding pipeline.
[153602](https://jira.devtools.intel.com/browse/CVS-153602)

Merge after: https://github.com/openvinotoolkit/openvino.genai/pull/1007/files

### CODE REVIEW

File: src/python/py_generate_pipeline.cpp [hunk 1]
```diff
@@ -763,6 +766,21 @@ PYBIND11_MODULE(py_generate_pipeline, m) {
...
+        .def(py::init([](
+            const std::string& model_path,
+            const std::string& device,
+            const ov::AnyMap& config,
+            const ov::genai::SchedulerConfig& scheduler_config
+        ) {
+            return ov::genai::draft_model(model_path, device, config, scheduler_config).second;
+        }),
+        py::arg("model_path"), "folder with openvino_model.xml and openvino_tokenizer[detokenizer].xml files", 
+        py::arg("device") = "CPU", "device on which inference will be done",
```
Comment (akuporos) [inline src/python/py_generate_pipeline.cpp:779]:
```suggestion
        py::arg("device") = "CPU", "device on which inference will be performed",
```

**Review by akuporos — APPROVED** [review]
Python part looks good for me

File: .github/workflows/causal_lm_cpp.yml [hunk 2]
```diff
@@ -432,6 +432,19 @@ jobs:
           assert predicted_greedy == predicted_speculative
           "
           echo "Alan Turing was a" passed
+      - name: run and compare python sample
+        run: |
+          source ./ov/setupvars.sh
+          && PYTHONPATH=./build/:$PYTHONPATH timeout 25s
```
Comment (ilya-lavrenov) [inline .github/workflows/causal_lm_cpp.yml:438]: use `env` statement similar to https://github.com/openvinotoolkit/openvino.genai/blob/c8ac7cb540753d03ca0966d84e46899f23a90dbf/.github/workflows/stable_diffusion_1_5_cpp.yml#L95

File: src/python/py_generate_pipeline.cpp [hunk 3]
```diff
@@ -769,6 +772,23 @@ PYBIND11_MODULE(py_generate_pipeline, m) {
             py::arg("streamer") = std::monostate{}
         );
     
+    py::class_<ov::Any>(m, "DraftModel", py::module_local(), "This class is used to enable Speculative Decoding")
```
Comment (ilya-lavrenov) [inline src/python/py_generate_pipeline.cpp:775]: not sure why in Python is called DraftModel, while in C++ draft_model ?

File: src/python/py_utils.hpp [hunk 4]
```diff
@@ -40,4 +40,6 @@ ov::genai::OptionalGenerationConfig update_config_from_kwargs(const ov::genai::O
 
 ov::genai::StreamerVariant pystreamer_to_streamer(const utils::PyBindStreamerVariant& py_streamer);
 
+ov::AnyMap kwargs_to_any_map(const py::kwargs& kwargs);
```
Comment (ilya-lavrenov) [inline src/python/py_utils.hpp:43]: it's already here

File: src/python/py_generate_pipeline.cpp [hunk 5]
```diff
@@ -768,10 +771,28 @@ PYBIND11_MODULE(py_generate_pipeline, m) {
...
+        py::arg("device") = "CPU", "device on which inference will be performed",
+        py::arg("config") = ov::AnyMap({}), "openvino.properties map",
+        py::arg("scheduler_config") = ov::genai::SchedulerConfig({}), "openvino.properties map"
+        );
+
+
     // init lora adapters
     init_lora_adapter(m);
-
+
```
Comment (ilya-lavrenov) [inline src/python/py_generate_pipeline.cpp:795]: please, drop leading spaces

File: samples/python/speculative_decoding_lm/speculative_decoding_lm.py [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+
+def main():
+    parser = argparse.ArgumentParser()
+    parser.add_argument('model_dir')
+    parser.add_argument('draft_model_dir')
+    parser.add_argument('prompt')
+    args = parser.parse_args()
+
+    # User can run main and draft model on different devices.
+    # Please, set device for main model in `LLMPipeline` constructor and in in `ov::genai::draft_model` for draft.
```
Comment (ilya-lavrenov) [inline samples/python/speculative_decoding_lm/speculative_decoding_lm.py:24]: `ov::genai::draft_model` is C++ entity. Please, use Python one

File: samples/CMakeLists.txt [hunk 7]
```diff
@@ -39,8 +39,8 @@ install(DIRECTORY
             python/greedy_causal_lm
             python/multinomial_causal_lm
             python/visual_language_chat
-            # python/speculative_decoding_lm
             python/text2image
             python/whisper_speech_recognition
+            python/speculative_decoding_lm
```
Comment (ilya-lavrenov) [inline samples/CMakeLists.txt:44]: samples are located here according to lexicographical order

**Review by ilya-lavrenov — APPROVED** [review]

File: src/python/py_generate_pipeline.cpp [hunk 8]
```diff
@@ -419,6 +420,23 @@ PYBIND11_MODULE(py_generate_pipeline, m) {
...
+            return std::make_unique<LLMPipeline>(model_path, device, utils::kwargs_to_any_map(kwargs));
+        }),
+        py::arg("model_path"), "folder with openvino_model.xml and openvino_tokenizer[detokenizer].xml files", 
+        py::arg("device") = "CPU", "device on which inference will be done",
+        R"(
+            LLMPipeline class constructor.
+            model_path (str): Path to the model file.
+            device (str): Device to run the model on (e.g., CPU, GPU). Default is 'CPU'.
+            Add {"scheduler_config": ov_genai.SchedulerConfig} to config properties to create continuous batching pipeline.
+        )")
```
Comment (ilya-lavrenov) [inline src/python/py_generate_pipeline.cpp:438]: @popovaan here several ctors for LLMPipeline has been added.
Please, consider merging them according to your suggestion with default args for `config` during your work on alignment of python ctors for GenAI pipeline

**Pull request merged**